lib/AddonSignals.py

```diff
@@ -75,33 +86,51 @@ def onNotification(self, sender, method, data):
         self._slots[sender][signal](_decodeData(data))
 
 
-class CallHandler:
-    def __init__(self, signal, data, source_id, timeout=1000):
+class CallHandler(xbmc.Monitor):
+    def __init__(self, signal, data, source_id, timeout=1000, use_timeout_exception=False):
+        xbmc.Monitor.__init__(self)
+        self.abort_requested = False
         self.signal = signal
         self.data = data
         self.timeout = timeout
         self.sourceID = source_id
         self._return = None
+        self.is_callback_received = False
+        self.use_timeout_exception = use_timeout_exception
         registerSlot(self.sourceID, '_return.{0}'.format(self.signal), self.callback)
         sendSignal(signal, data, self.sourceID)
 
+    def onNotification(self, sender, method, data):
+        # Workaround to know when Kodi will be terminated, the method Monitor().abortRequested() not works on scripts
+        if method in ['System.OnQuit', 'System.OnRestart', 'System.OnSleep']:
+            self.abort_requested = True
```

Comment:
Curious why `self.abortRequested()` (since this is a Monitor class) wouldn't work.

Reply:
hi ruuk
I did some tests, although i didn't understand the reason,
these two commands (abortRequested/waitForAbort) only works within an addon service instance

Reply:
Did you try `self.abortRequested()` vs. `Monitor().abortRequested()`? In the past I've found issues with multiple Monitor instances. I use abortRequested() extensively in the Plex add-on (using a single Monitor instance) and haven't had issues.

Reply:
i can try, but this means create hundred..thousand.. and over of object Monitor instances throughout the waiting time
not seem to me a good thing to do

Reply:
I'm not sure what you mean. `self.abortRequested()` shouldn't create any Monitor instances. `self` is a Monitor instance in this case.

Reply:
maybe i misunderstood?
you have write this `Monitor().abortRequested()`
this `Monitor()` means create a instance of monitor in same place

if you do something like:

<pre>
while not self.is_callback_received:
   if Monitor().abortRequested():
      ....
</pre>

means create and destroy every 10ms thousand and over of monitor objects throughout the timeout
that's why i said it's not a good behaviour,
at least the verification of a boolean variable (assegned by event callback) does not require any effort of the cpu

instead, if you meant using self.abortRequested() to verify kodi close, already tried.. not works

I'm open to try other solutions, if any

Reply:
Just to be clear, you tried `self.abortRequested()` in this addon before you settled on the notification method? I just want to verify, because I've had issues with creating multiple instances on `xbmc.Monitor` in an addon, but have never seen an issue if I made sure I had only instance of `xbmc.Monitor` for the whole app.

For example, in the Plex add-on, we create one instance of `xbmc.Monitor` that is used globally throughout the app for all `abortRequested` calls.

Reply:
hi
well seem i found the issue,
i'm working on windows, and usually when i want to stop a operation in progress, i close kodi with the "X" button on windows title bar or also i use the kodi shutdown menu (S key)

so when you use these the monitor abort not works...
https://forum.kodi.tv/showthread.php?tid=355930&pid=2964629#pid2964629

only if you press back key on loading screen will be aborted correctly
that's why I couldn't get it to work

i think there is a bug in kodi core
next i will try other test

Reply:
i confirm is working, then i have fixed the commit
Unfortunately the methods i used to stop kodi are not managed by kodi and so i opened an Issue on Kodi side